Makefile

```
@@ -106,58 +106,68 @@ $(K0S_GO_BUILD_CACHE):
+controllergen_targets += pkg/apis/autopilot/v1beta2/.controller-gen.stamp
 pkg/apis/autopilot/v1beta2/.controller-gen.stamp: $(shell find pkg/apis/autopilot/v1beta2/ -maxdepth 1 -type f -name \*.go)
 pkg/apis/autopilot/v1beta2/.controller-gen.stamp: gen_output_dir = autopilot
 
+codegen_targets += $(controllergen_targets)
 pkg/apis/%/.controller-gen.stamp: .k0sbuild.docker-image.k0s hack/tools/boilerplate.go.txt hack/tools/Makefile.variables
 	rm -rf 'static/manifests/$(gen_output_dir)/CustomResourceDefinition'
-	rm -f -- '$(dir $@)'zz_*.go
-	CGO_ENABLED=0 $(GO) install sigs.k8s.io/controller-tools/cmd/controller-gen@v$(controller-gen_version)
-	$(GO_ENV) controller-gen \
-	  crd \
-	  paths="./$(dir $@)..." \
-	  output:crd:artifacts:config=./static/manifests/$(gen_output_dir)/CustomResourceDefinition \
-	  object:headerFile=hack/tools/boilerplate.go.txt
+	mkdir -p 'static/manifests/$(gen_output_dir)'
+	gendir="$$(mktemp -d .controller-gen.XXXXXX.tmp)" \
+	  && trap "rm -rf -- $$gendir" INT EXIT \
+	  && CGO_ENABLED=0 $(GO) run sigs.k8s.io/controller-tools/cmd/controller-gen@v$(controller-gen_version) \
+	    paths="./$(dir $@)..." \
+	    object:headerFile=hack/tools/boilerplate.go.txt output:object:dir="$$gendir" \
+	    crd output:crd:dir='static/manifests/$(gen_output_dir)/CustomResourceDefinition' \
+	  && mv -f -- "$$gendir"/zz_generated.deepcopy.go '$(dir $@).'
 	touch -- '$@'
 
 clientset_input_dirs := pkg/apis/autopilot/v1beta2 pkg/apis/k0s/v1beta1
 codegen_targets += pkg/client/clientset/.client-gen.stamp
 pkg/client/clientset/.client-gen.stamp: $(shell find $(clientset_input_dirs) -type f -name \*.go -not -name \*_test.go -not -name zz_\*)
 pkg/client/clientset/.client-gen.stamp: .k0sbuild.docker-image.k0s hack/tools/boilerplate.go.txt embedded-bins/Makefile.variables
-	rm -rf -- '$(dir $@)'
-	CGO_ENABLED=0 $(GO) run k8s.io/code-generator/cmd/client-gen@v$(kubernetes_version:1.%=0.%) \
-	  --go-header-file=hack/tools/boilerplate.go.txt \
-	  --input-base='' \
-	  --input=$(subst $(space),$(comma),$(clientset_input_dirs:%=github.com/k0sproject/k0s/%)) \
-	  --output-package=github.com/k0sproject/k0s/$(patsubst %/,%,$(dir $(patsubst %/,%,$(dir $@)))) \
-	  --clientset-name=$(notdir $(patsubst %/,%,$(dir $@))) \
-	  --output-base=. \
-	  --trim-path-prefix=github.com/k0sproject/k0s
+	gendir="$$(mktemp -d .client-gen.XXXXXX.tmp)" \
+	  && trap "rm -rf -- $$gendir" INT EXIT \
+	  && CGO_ENABLED=0 $(GO) run k8s.io/code-generator/cmd/client-gen@v$(kubernetes_version:1.%=0.%) \
+	    --go-header-file=hack/tools/boilerplate.go.txt \
+	    --input-base='' \
+	    --input=$(subst $(space),$(comma),$(clientset_input_dirs:%=github.com/k0sproject/k0s/%)) \
+	    --output-package=github.com/k0sproject/k0s/pkg/client \
+	    --clientset-name=clientset \
+	    --output-base="$$gendir/out" \
+	    --trim-path-prefix=github.com/k0sproject/k0s \
+	  && { [ ! -e pkg/client/clientset ] || mv -- pkg/client/clientset "$$gendir/old"; } \
```

What is the reasoning behind this? As I read how this works:

- if pkg/client/clientset exists, move it to temp directory and delete it directly with the INT/EXIT trap.

Why not delete it immediatly and ignore errors? (to avoid race condition?)

```suggestion
	  && { mv -- pkg/client/clientset "$$gendir/old" 2>/dev/null || true; } \
```

It is a general good idea to avoid test if something exists before operation, as it is theoretically possible that the thing is created/removed between the test and the operation. Better to just do the operation and deal with (or ignore) the not-found error.

---

> Better to just do the operation and deal with (or ignore) the not-found error.

Agree.  However, with `mv`, there's no way to distinguish the not found case from any other error. Redirecting that to /dev/null even doesn't allow for spotting anything in the logs after the fact. I'd prefer to leave it as is, so we actually fail if `mv` failed. Even if the check is racy, nothing should remove that directory in the meantime anyways, hence I'd prefer that operation to fail in that case.